sdks/python/apache_beam/io/filesystem.py

```diff
@@ -250,8 +255,73 @@ def flush(self):
 
   @property
   def seekable(self):
-    # TODO: Add support for seeking to a file position.
-    return False
+    return self._file.mode == 'r'
+
+  def _rewind_read_buffer(self):
+    self._read_position = 0
+    self._read_buffer.seek(0)
+    self._read_buffer.truncate(0)
+
+  def _rewind_file(self):
+    self._file.seek(0, os.SEEK_SET)
+    self._read_eof = False
+    self._uncompressed_position = 0
+
+  def _rewind(self):
+    self._rewind_read_buffer()
+    self._rewind_file()
+
+    # Re-initialize decompressor to clear any data buffered prior to rewind
+    self._initialize_decompressor()
+
+  def seek(self, offset, whence=os.SEEK_SET):
+    """Set the file's current offset.
+
+    Seeking behavior:
+      * seeking from the end (SEEK_END) the whole file is decompressed once to
+        determine it's size. Therefore it is preferred to use
+        SEEK_SET or SEEK_CUR to avoid the processing overhead
+      * seeking backwards from the current position rewinds the file to 0
+        and decompresses the chunks to the requested offset
+      * seeking is only supported in files opened for reading
+      * if the new offset is out of bound, it is adjusted to either 0 or EOF.
+
+    Args:
+      offset: seek offset as number.
+      whence: seek mode. Supported modes are os.SEEK_SET (absolute seek),
+        os.SEEK_CUR (seek relative to the current position), and os.SEEK_END
+        (seek relative to the end, offset should be negative).
+
+    Raises:
+      IOError: When this buffer is closed.
+      ValueError: When whence is invalid or the file is not seekable
```

Comment:
Out of bound seek is explained in 'seeking behavior' section in this docstring. I'd prefer to keep it there as the behavior does not yield to exceptions.